### PR TITLE
Dala 3045 manual maintenance sprint 57

### PR DESCRIPTION
**Pull request by HussainRoker**
# Manual Maintenance Sprint [57]

Note: To create a PR using this template add the query parameter `template=manual_maintenance.md` to the merge request
creation URL (or simply copy this md file into the description)

# Maintenance tasks (to be completed by the assignee)

### Notes from Dec 2023
- Timeout not working
- There were two 2 messages found in RabbitMQ dead letter queue on dev1 and dev2 respectively, details added in ticket
- JDK/JRE must remain at 17, upgrading it to 21 caused too many errors, also Kotlin 1.9.10 is not compatible with Java 21
- kotlin to remain at 1.9.10 instead of 1.9.20, which caused error
- eclipse-temurin exists in version 21 already but cannot be updated, as that breaks e2e tests in CI, we're using 17.
- some sec fixes or updates to `package.json` in /frontend and /keycloak break the build

### Skipped updates

The following known issues need to be reviewed in case a compatible version is available. Add new known issues as they
appear.

- [ ] Update keycloak to >=22.0.3 is skipped due to failing all e2e tests
- [ ] Update e2etests/Dockerfile update breaks the build
- [ ] Update "@vue/tsconfig" to >=0.2.0 introduces major changes in typescript rules (~500 TS Errors throughout the
  project and unresolved imports that are hard to fix), skipped.
- [ ] Update Cypress to >= 12.12.0 introduces an issue with the usage of `$route` in component test 
  `DatasetOverview.cy.ts`. Issue with Cypress has been created to hopefully resolve this from the side of Cypress:
  https://github.com/cypress-io/cypress/issues/26902
- [ ] Update Ktlint to >= 49.0 breaks the ktlint tasks (issue described here: 
  https://github.com/JLLeitschuh/ktlint-gradle/issues/665 and possible fix here: 
  https://github.com/JLLeitschuh/ktlint-gradle/pull/667)
- [ ] Update @zxcvbn-ts/language-common to 3.0.3 is skipped due to issues in rebuilding keycloak Docker images
- [ ] Update @zxcvbn-ts/language-en to 3.0.1 is skipped due to issues in rebuilding keycloak Docker images
- [ ] The docker-compose-plugin v.2.19.1 causes connection issues:
- [ ] Check that it is still valid for `**/CompanyApi.kt', '**/CompanyDataController.kt` to be excluded from detekt, 
      at latest once the refactoring of the APIs is done this must be reevaluated
### Gradle update

- [ ] Execute `gradlew dependencyUpdates` to get a report on Dependencies with updates
- [ ] Execute `refreshVersions` in Gradle tasks or `gradlew refreshVersions` to generate version suggestions in `versions.properties`
- [ ] Update versions in `versions.properties`
- [ ] Update the gradle wrapper: execute `gradle wrapper --gradle-version X.Y.Z`

### Dataland frontend

- [ ] Update node version in `dataland-frontend/build.gradle.kts`
- [ ] Update node packages: run the `updatepackages` script, e.g. by  `npm run updatepackages` to update versions in
  package.json
- [ ] Run the `updatepackagelock`, e.g. by  `npm run updatepackagelock` script to update `package-lock.json` and check
  for security issues

### Dataland keycloak theme

- [ ] Update node version in `dataland-keycloak/dataland_theme/login/build.gradle.kts`
- [ ] Update node packages: run the `updatepackages` script, e.g. by  `npm run updatepackages` to update versions in
  package.json
- [ ] Run the `updatepackagelock`, e.g. by  `npm run updatepackagelock` script to update `package-lock.json` and check
  for security issues

### Dockerfile updates

Update versions in the following dockerfiles

- [ ] `./dataland-api-key-manager/Dockerfile`
- [ ] `./dataland-api-key-manager/DockerfileBase`
- [ ] `./dataland-api-key-manager/DockerfileTest`
- [ ] `./dataland-backend/Dockerfile`
- [ ] `./dataland-backend/DockerfileBase`
- [ ] `./dataland-backend/DockerfileTest`
- [ ] `./dataland-document-manager/Dockerfile`
- [ ] `./dataland-document-manager/DockerfileBase`
- [ ] `./dataland-document-manager/DockerfileTest`
- [ ] `./dataland-e2etests/Dockerfile`
- [ ] `./dataland-e2etests/DockerfileBase`
- [ ] `./dataland-frontend/Dockerfile`
- [ ] `./dataland-frontend/DockerfileTest`
- [ ] `./dataland-internal-storage/Dockerfile`
- [ ] `./dataland-internal-storage/DockerfileBase`
- [ ] `./dataland-internal-storage/DockerfileTest`
- [ ] `./dataland-qa-service/Dockerfile`
- [ ] `./dataland-qa-service/DockerfileBase`
- [ ] `./dataland-qa-service/DockerfileTest`
- [ ] `./dataland-rabbitmq/Dockerfile`
- [ ] `./dataland-inbound-admin-proxy/Dockerfile`
- [ ] `./dataland-inbound-proxy/Dockerfile`
- [ ] `./dataland-pgadmin/Dockerfile`
- [ ] `./dataland-keycloak/Dockerfile`  (also update realm json files with new version)
- [ ] `./base-dockerfiles/DockerfileGradle`
- [ ] Update the versions of the external images for api-key-manager-db, backend-db, keycloak-db, internal-storage-db,
  document-manager-db, qa-service-db, community-manager-db and frontend-dev in `./docker-compose.yml`
- [ ] Check if there are any services in the `docker-compose.yml` file that have not gotten an update yet (e.g. a new
  service that is not covered by the tasks above)

## Dataland Monitoring

- [ ] Go to the monitoring repository and execute the tasks described in the manual maintenance template there

## Server updates

Execute `sudo apt-get update && sudo apt-get upgrade` on

- [ ] dev1.dataland.com
- [ ] dev2.dataland.com
- [ ] test.dataland.com
- [ ] letsencrypt.dataland.com
- [ ] monitoring.dataland.com
- [ ] (OPT) dataland.com

### ssh-keys maintenance

check that all ssh-keys are set and erased from people that have left

- [ ] dev1.dataland.com
- [ ] dev2.dataland.com
- [ ] test.dataland.com
- [ ] letsencrypt.dataland.com
- [ ] monitoring.dataland.com
- [ ] (OPT) dataland.com

## Check keycloak automatic logout if inactive

- [ ] Check that you are automatically logged out if you are idle for too long and also get notified about this by a
  pop-up. Timeouts are defined as `TIME_BEFORE_REFRESH_TOKEN_EXPIRY_TO_DISPLAY_SESSION_WARNING_IN_MS`
  in `./dataland-frontend/src/utils/Constants.ts` and `ssoSessionIdleTimeout`
  in `./dataland-keycloak/realms/datalandsecurity-realm.json`.

## Check e-mails sent by keycloak

- [ ] Check that the verification e-mail is sent and displayed correctly
- [ ] Check that the reset password e-mail is sent and displayed correctly
- [ ] Check that the password can be reset
- [ ] Check that account linking via e-mail verification works correctly
- [ ] Check that account linking via username & password verification works correctly

## Check e-mails sent by backend

- [ ] Send an invitation (data) request from one of the dev servers and check if the e-mail to dataland@d-fine.com
  contains the right attachments and is displayed correctly.

## Check RabbitMQ dead letter queue and disk space

- [ ] RabbitMQ does need at least 768MB of free disk space to operate. `ssh` into all servers and check the available
  disk space with `df` command. If the open disk space is close to the minimum requirement, clear up disk space
  with `sudo docker image prune --all`.
- [ ] On all environments, no new messages should have been added to the dead letter queue since the last manual
  maintenance. If new messages have appeared this does need to be investigated. The dead letter queue can be accessed
  and messages on it read in the RabbitMQ GUI. Access it by port-forwarding port `6789` from the server and then
  accessing the GUI at `localhost:6789/rabbitmq`. After login, the dead letter queue can be found at Queues &rarr;
  deadLetterQueue &rarr; Get message.

## Check that links to external webpages are working
- [ ] Go to the frameworks/upload page and check that the link to WWF Pathway to Paris is still working as intended.

## Check that the main branch has no sonar issues
- [ ] Go to the sonar report summary of the main branch and verify that there are no sonar findings. If there are sonar 
  findings fix them.

## Conclusion

- [ ] After updating all components check if everything still works
- [ ] The new version is deployed to a dev server using this branch
    - [ ] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
    - [ ] It's verified that everything seems to be working fine by manually using the website
    - [ ] All implemented Social Logins have been tested manually in the UI
- [ ] This template has been updated to reflect the latest state of tasks required and known issues with upgrades
- [ ] The Merge Request commit message needs to contain 'manual maintenance' to satisfy the CI maintenance check in
  future commits

# Review (to be completed by the reviewer)

- [x] The Github Actions (including Sonarqube Gateway and Lint Checks) are green. This is enforced by Github.
- [x] A peer-review has been executed
  - [x] The code has been manually inspected by someone who did not implement the feature
- [x] The PR actually implements what is described above
- [x] Documentation is updated as required
- [x] The automated deployment is updated if required
- [x] The new version is deployed to the dev server "dev1" using this branch
  - [x] Run with setting `Reset non-user related Docker Volumes & Re-populate` turned on
  - [x] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
  - [x] It's verified that the CD run is green
  - [x] It's verified that everything seems to be working fine by manually using the website
  - [x] All implemented Social Logins have been tested manually in the UI
  - [x] The "mailto:*" buttons ("I am interested" and "Get in touch") on the landing page work.
  - [x] Go to the swagger-UI, authorize, run a "GET" request to the companies endpoint and assure that your
    authorization has worked by assuring that you get a 200 response
- [x] If any work on the UI is to be merged, those changes were also documented in the Figma
- [x] The local Dev stack still works: execute `startDevelopmentStack.sh` and execute `npm run testpipeline`
  and `npm run testcomponent` in dataland-frontend
- [x] After(!) the cypress tests have passed locally, execute the backend-e2e-tests `./gradlew dataland-e2etests:test`
- [x] Locally: Go to the swagger-UI, authorize, run a "GET" request to the companies endpoint and assure that your
  authorization has worked by assuring that you get a 200 response
- [x] Merge using Squash Commit. The Merge Commit Message needs to contain "Manual Maintenance"
- [ ] After merge check SonarQube state of main branch at https://sonarcloud.io/summary/new_code?id=d-fine_Dataland. 
  The full scan might reveal new issues (e.g. deprecation) on old code which is generally not detected on the branch.

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md
Comment (aliabdoli77) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:1]: "Notes from month X" is an unnecessary section
Updating issues should be integrated into the known issues section. We don't need to explain in detail how or why it failed. Things like RabbitMQ error messages should be inspected and discussed in the team chat if necessary.
@HussainRoker

Comment (aliabdoli77) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:1]: I think "Notes from X" was added since the last MM. It isn't originally a section in the MM checklist

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md [hunk 1]
```diff
@@ -7,18 +7,17 @@ creation URL (or simply copy this md file into the description)
...
 
 The following known issues need to be reviewed in case a compatible version is available. Add new known issues as they
 appear.
 
+- [ ] postgres can't be updated to 16 as existing data in dev instances are not compatible
+- [ ] There were two 2 messages found in RabbitMQ dead letter queue on dev1 and dev2 respectively, details added in ticket
+- [ ] JDK/JRE must remain at 17, upgrading it to 21 caused too many errors, also Kotlin 1.9.10 is not compatible with Java 21
+- [ ] kotlin to remain at 1.9.10 instead of 1.9.20, which caused error
+- [ ] eclipse-temurin exists in version 21 already but cannot be updated, as that breaks e2e tests in CI, we're using 17.
+- [ ] some sec fixes or updates to `package.json` in /frontend and /keycloak break the build
```
Comment (fsoffel) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:20]: What are sec fixes? Secondary version upgrades?

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md [hunk 2]
```diff
@@ -31,10 +30,7 @@ appear.
   https://github.com/JLLeitschuh/ktlint-gradle/pull/667)
 - [ ] Update @zxcvbn-ts/language-common to 3.0.3 is skipped due to issues in rebuilding keycloak Docker images
 - [ ] Update @zxcvbn-ts/language-en to 3.0.1 is skipped due to issues in rebuilding keycloak Docker images
-- [ ] Update Postgres in Docker-compose.yml to 16.0 causes CD to fail. Should be retried in future tickets.
 - [ ] The docker-compose-plugin v.2.19.1 causes connection issues:
-  If running `sudo apt-get update && sudo apt-get upgrade` on the servers causes connection issues
```
Comment (fsoffel) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:36]: Why was this removed

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md (hunk 1 above)
Comment (HussainRoker) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:20]: I guess security fixes, in keycloak have not encountered such @PaulChr99  has worked on frontend,

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md (hunk 2 above)
Comment (HussainRoker) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:36]: my bad, resolved

Comment (fsoffel) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:36]: `If running `sudo apt-get update && sudo apt-get upgrade` on the servers causes connection issues...` This is what I was referring to

Comment (HussainRoker) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:36]: I didn't got any

Comment (fsoffel) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:36]: Thinking about it. Fair enough.
